<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SingleCell RNAseq templates â€¢ hbcABC</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="SingleCell RNAseq templates">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hbcABC</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/general_start.html">General templates</a>
    </li>
    <li>
      <a href="../articles/how_to_add_templates.html">How to create templates</a>
    </li>
    <li>
      <a href="../articles/singlecell-tutorial.html">SingleCell RNAseq templates</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>SingleCell RNAseq templates</h1>
                        <h4 class="author">Lorena Pantano</h4>
            
      
      
      <div class="hidden name"><code>singlecell-tutorial.rmd</code></div>

    </div>

    
    
<div id="single-cell-templates" class="section level1">
<h1 class="hasAnchor">
<a href="#single-cell-templates" class="anchor"></a>Single cell templates</h1>
<p>Please, contribute with your improvements. I am sure you have something to add.</p>
<div id="qc" class="section level2">
<h2 class="hasAnchor">
<a href="#qc" class="anchor"></a>QC</h2>
<p><code>simple_qc_single_cell</code> is a template to run a fast QC with ggplot code to modify the plots as needed.</p>
<p>To create the template, use the Rstudio menu as for any other template or:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">source =<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">system.file</span>(<span class="st">"rmarkdown"</span>, <span class="dt">package=</span><span class="st">"hbcABC"</span>),
          <span class="st">"templates"</span>, <span class="st">"sinple_qc_single_cell"</span>,
          <span class="st">"skeleton"</span>,
          <span class="st">"skeleton.rmd"</span>)
<span class="kw">file.copy</span>(source, <span class="st">"reports/simple_qc_sc.rmd"</span>)</code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<p>The next step is to run the normalization, scaling and clustering that are high demand in computation time and memory.</p>
</div>
<div id="clustering-with-seurat" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering-with-seurat" class="anchor"></a>Clustering with Seurat</h2>
<p>There is one script that can be used to run all the main steps in the cluster and save the most important objects to be loaded in the Rmd for clustering visualization.</p>
<p>To get the <a href="https://github.com/hbc/hbcABC/tree/master/inst/rmarkdown/Rscripts/singlecell/clustering_seurat.R">script</a> for this step, do:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">source =<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">system.file</span>(<span class="st">"rmarkdown"</span>, <span class="dt">package=</span><span class="st">"hbcABC"</span>),
          <span class="st">"Rscripts"</span>, 
          <span class="st">"singlecell"</span>,
          <span class="st">"clustering_seurat.R"</span>)
<span class="kw">file.copy</span>(source, <span class="st">"clustering.r"</span>)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>This script uses the object saved in <code>simple_qc_single_cell.rmd</code>:</p>
<ul>
<li>seurat_raw.rds</li>
<li>cycle.rda</li>
</ul>
<p>There is an important parameter in the script, <code>pcs</code>. It is calculated automatically as Michael Steinbaugh with some modifications from Lorena Pantano. It is important to look at this figure to check whether the value makes sense or not.</p>
<p>The idea behind this is to choose the value just at the elbow of the plot. It helps to give to clients the list of top genes for each PCs so they can help to define the cutoff based on prior information.</p>
<p>This script will create the following objects:</p>
<ul>
<li>seurat_pre_regress.rds</li>
<li>seurat_tsne.rds</li>
</ul>
<p>These objects will be used by <code>seurat_clustering_single_cell</code> to perform best practices as described by seurat tutorial.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">source =<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">system.file</span>(<span class="st">"rmarkdown"</span>, <span class="dt">package=</span><span class="st">"hbcABC"</span>),
          <span class="st">"templates"</span>, <span class="st">"seurat_clustering_single_cell"</span>,
          <span class="st">"skeleton"</span>,
          <span class="st">"skeleton.rmd"</span>)
<span class="kw">file.copy</span>(source, <span class="st">"reports/seurat_clustering.rmd"</span>)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="find-markers-with-seurat" class="section level2">
<h2 class="hasAnchor">
<a href="#find-markers-with-seurat" class="anchor"></a>Find Markers with Seurat</h2>
<p>When you are happy with your clusters you can jump into the markers analysis. Running <code>FindAllMarkers</code> function (with Wilcox) can lead to false positive but it should give an idea of the markers for each cluster.</p>
<p>One can use other methods, like differential expression type, comparing one clusters against all. Right now, if the project has big number of cells &gt; 10000, other method will need high amount of memory and time and it is difficult to handle, so extra hours should be schedule for this.</p>
<p>To use the template <a href="https://github.com/hbc/hbcABC/tree/master/inst/rmarkdown/Rscripts/singlecell/markers_seurat.R">script</a> for this analysis:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">source =<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">system.file</span>(<span class="st">"rmarkdown"</span>, <span class="dt">package=</span><span class="st">"hbcABC"</span>),
          <span class="st">"Rscripts"</span>,
          <span class="st">"singlecell"</span>,
          <span class="st">"markers_seurat.R"</span>)

<span class="kw">file.copy</span>(source, <span class="st">"markers_seurat.r"</span>)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="subset-of-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#subset-of-clusters" class="anchor"></a>Subset of clusters</h2>
<p>Sometimes we are interested on running Seurat for a subset of cells.</p>
<p>This <a href="https://github.com/hbc/hbcABC/tree/master/inst/rmarkdown/Rscripts/singlecell/clustering_seurat_subset.R">script</a> help with that:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">source =<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">system.file</span>(<span class="st">"rmarkdown"</span>, <span class="dt">package=</span><span class="st">"hbcABC"</span>),
          <span class="st">"Rscripts"</span>,
          <span class="st">"singlecell"</span>,
          <span class="st">"clustering_seurat_subset.R"</span>)

<span class="kw">file.copy</span>(source, <span class="st">"subset_seurat.r"</span>)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="differential-expression-among-groups-for-each-cluster" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression-among-groups-for-each-cluster" class="anchor"></a>Differential expression among groups for each cluster</h2>
<p>In the case the project has different groups of cells, one can run differential expression analysis for each cluster. For instance, imagine you have two groups: A and B, and you want to know the differences between them in each single cluster. One option is to use <a href="https://bioconductor.org/packages/release/bioc/html/zinbwave.html">zinbwave</a> - <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a> combo to account for the zero-inflation distribution in single cell.</p>
<p>This <a href="https://github.com/hbc/hbcABC/tree/master/inst/rmarkdown/Rscripts/singlecell/de_with_zinbwave.R">script</a> can help with that:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">source =<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">system.file</span>(<span class="st">"rmarkdown"</span>, <span class="dt">package=</span><span class="st">"hbcABC"</span>),
          <span class="st">"Rscripts"</span>,
          <span class="st">"singlecell"</span>,
          <span class="st">"de_with_zinbwave.R"</span>)

<span class="kw">file.copy</span>(source, <span class="st">"de_with_zinbwave.r"</span>)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#single-cell-templates">Single cell templates</a><ul class="nav nav-pills nav-stacked">
<li><a href="#qc">QC</a></li>
      <li><a href="#clustering-with-seurat">Clustering with Seurat</a></li>
      <li><a href="#find-markers-with-seurat">Find Markers with Seurat</a></li>
      <li><a href="#subset-of-clusters">Subset of clusters</a></li>
      <li><a href="#differential-expression-among-groups-for-each-cluster">Differential expression among groups for each cluster</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Lorena Pantano, Michael Steinbaugh, Rory Kirchner, Victor Barrera, Mary Piper, Meeta Mistry, Radhika Khetani, Kayleigh Rutherford, Oliver Hofmann, John Hutchinson, Shannan Ho Sui, Harvard Chan Bioinformatics Core.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
